correctly marshal constant arrays

IIFE opened the pull request.

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/mono/CppSharp?pullRequest=1346) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/mono/CppSharp?pullRequest=1346) before we can accept your contribution.<br/><br/>:x: IIFE [sign now](https://cla.dotnetfoundation.org/mono/CppSharp?pullRequest=1346)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/mono/CppSharp?pullRequest=1346) it.</sub>

ddobrev commented:
@IIFE I've merged one of your PR-s and there conflicts now, please resolve.

IIFE commented:
@ddobrev @tritao I've updated this PR with latest changes from master.

I've purposely only put tests for CLI because C# generator appears to have issues with the test cases. Obviously it needs to be looked at and fixed, but probably best done as a separate PR so that we're not holding off the CLI fixes.

Thanks.

IIFE commented:
@tritao @ddobrev Would it be possible to progress this PR please? Let me know if you have any questions.

Thanks.

tritao commented on src/Generator/Generators/CLI/CLIMarshal.cs:
```diff
@@ -789,23 +793,26 @@ private void MarshalValueClassProperty(Property property, string marshalVar)
-                !type.IsPrimitiveType();
-
-            if (isRef)
+            if (!string.IsNullOrWhiteSpace(marshal.Context.Return))
```
Why did you add this check here?

It is my understanding that the `Return` here should always provide a valid value upon invoking the marshaler. If not, then it should an error, no?

IIFE replied:
Hi. This is for the scenario where we are marshalling an array member of a value/struct type. The array member is set and marshalled inside VisitArrayType on line 426. There is no return involved in that case, so we should not set a return here.

Basically without the if you end up with code like this:
```
    auto _marshal0 = ::UnionNestedInsideStruct();
    if (value.SzText != nullptr)
    {
        if (value.SzText->Length != 10)
        {
            throw gcnew System::InvalidOperationException("Source array size must equal destination array size.");
        }
        for (int i = 0; i < 10; i++)
            _marshal0.szText[i] = value.SzText[i];
    }
    _marshal0.szText = ;
    ((::StructWithNestedUnion*)NativePtr)->nestedUnion = _marshal0;
```

Notice the `    _marshal0.szText = ;`.

tritao commented:
> @tritao @ddobrev Would it be possible to progress this PR please? Let me know if you have any questions.
> 
> Thanks.

Sorry for the delay in processing this one.

Looks mostly good, I have just one question I left as a review.

IIFE commented:
@tritao No problem. Hopefully my answer to your question makes sense.

tritao commented:
Thanks again for the contribution.
